sdk/client/client.ts

```diff
@@ -124,49 +123,146 @@ export class Client {
+
+  private configurationMapping: Array<Array<number|Array<string>>> = [
+    [
+      Source.Environment,
+      [
+        'FBN_TOKEN_URL',
+        'FBN_USERNAME',
+        'FBN_PASSWORD',
+        'FBN_CLIENT_ID',
+        'FBN_CLIENT_SECRET',
+        'FBN_LUSID_API_URL'
+      ]
+    ],
+    [
+      Source.Secrets,
+      [
+        'tokenUrl',
+        'username',
+        'password',
+        'clientId',
+        'clientSecret',
+        'apiUrl'
+      ]
+    ]
+  ]
+
   // The refresh limit in seconds before token expiry to trigger a refresh
-  refreshLimit: number = refreshLimit
+  private refreshLimit: number = 3580
+
+  private loadSecretsFile( secretsFilePath?: string ): object {
+
+    try
+    {
+      return this.secretsFileContent = require( !!secretsFilePath ? secretsFilePath : this.configuration.secretsFilePath ).api;
+    }
+    catch( e )
+    {
+      if( e.code === 'MODULE_NOT_FOUND' )
+      {
+        if( !!secretsFilePath )
+        {
+          // a specific path was requested
+
+          throw `Configuration file ${secretsFilePath} was not found`;
+        }
+
+        // no specific path requested, ergo it went with the default path
+        // a secrets file not found is not necessarily a problem
+        // this will reset the internal validation to an empty configuration
+
+        return this.secretsFileContent = {};
+      }
+
+      // module was found but there's some other (random) error
+
+      throw e;
+    }
+
+  }
 
   // Constructor method which takes the details on where to find the credentials
   constructor(
-    tokenUrlDetails: [Source, string],
-    usernameDetails: [Source, string],
-    passwordDetails: [Source, string],
-    clientIdDetails: [Source, string],
-    clientSecretDetails: [Source, string],
-    apiUrlDetails: [Source, string],
+      {
+        tokenUrlDetails,
+        usernameDetails,
+        passwordDetails,
+        clientIdDetails,
+        clientSecretDetails,
+        apiUrlDetails,
+        secretsFilePath,
+      }: {
+        tokenUrlDetails?: string,
+        usernameDetails?: string,
+        passwordDetails?: string,
+        clientIdDetails?: string,
+        clientSecretDetails?: string,
+        apiUrlDetails?: string,
+        secretsFilePath?: string
+      } = {
+        // empty object by default, thus triggering an auto failover
+      }
     ) {
 
-    // Set the path to the secrets file
-    this.secretsFilePath = secretsPath
+    this.configuration.secretsFilePath = !!secretsFilePath
+      ? secretsFilePath
+      : this.configuration.secretsFilePath
+
+    // attempt to get disk config by default. if it doesn't exist it is still fine
+    this.loadSecretsFile( secretsFilePath );
+
+    // provide init + default to environment vars below
+    this.configuration.tokenUrlDetails = !!tokenUrlDetails
+      ? this.fetchConfigurationItem( Source.Raw, tokenUrlDetails, true )
+      : this.fetchConfigurationItem( Source.Secrets, 'tokenUrl', false );
```

Comment:
In this instance, if I provide in tokenUrlDetails when creating my client 'tokenUrl' then won't this actually use the value of 'tokenUrl' as my tokenUrl?